enclave/impl/src/message_bus/mod.rs

```diff
@@ -0,0 +1,753 @@
+        &mut self,
+        query: &QueryRequest,
+        challenge_value: &[u8; 32],
+    ) -> Result<QueryResponse, QueryError> {
+        // First, authenticate the user, checking their signature over the
+        // challenge value.
+        let decompressed_identity = RistrettoPublic::try_from(&query.auth_identity[..])?;
+        let sig = RistrettoSignature::try_from(&query.auth_signature[..])?;
+        decompressed_identity.verify_schnorrkel(
+            GRAPEVINE_CHALLENGE_SIGNING_CONTEXT,
+            challenge_value,
+            &sig,
+        )?;
+
+        // In the specific case that the caller did not set request type to a nonzero
+        // value, fail hard and fast, because this breaks security.
+        // Otherwise, I believe that unknown request types will be interpretted as READ
+        if bool::from(query.request_type.ct_eq(&0)) {
+            return Err(QueryError::InvalidRequestType);
+        }
+
+        // Call the appropriate helper
+        match query.request_type {
+            REQUEST_TYPE_CREATE => Ok(self.create_record(query)?),
+            _ => Ok(self.access_record(query)?),
+        }
+    }
+
+    // Handle a create record request.
+    fn create_record(&mut self, query: &QueryRequest) -> Result<QueryResponse, QueryError> {
+        let mut rng = McRng::default();
+        let (mut id, mut marshalled) =
+            <A8Bytes<MessageBusValueSize> as MarshalledRecord>::new_from_proto(&query.record)?;
+
+        // Randomize the id to a nonzero value no matter what the user specified.
+        // This is good because it prevents them from probing for specific pre-existing ids
+        // in order to see when they may be created or deleted.
+        rng.fill_bytes(id.as_mut_slice());
+        while bool::from(id.ct_eq(&Default::default())) {
+            rng.fill_bytes(id.as_mut_slice());
+        }
+
+        // Set the timestamp to our value
+        *marshalled.get_timestamp_mut() = self.current_timestamp;
+        // Set the sender identity to be the authorized identity value.
+        marshalled
+            .get_sender_mut()
+            .copy_from_slice(&query.auth_identity);
+
+        let mut insertion_succeeds = Choice::from(0);
+        let mut message_omap_result_code = 0u32;
+        let recipient_omap_result_code = self.recipients_to_ids.access_and_insert(
+            marshalled.get_recipient(),
+            &Default::default(),
+            &mut rng,
+            |_code, buffer| {
+                let mut found_key = Choice::from(0);
+                let mut found_key_needs_overwrite = Choice::from(0);
+                let mut found_free_space = Choice::from(0);
+                let mut desired_index = u64::MAX;
+
+                // Find a place where we can write a new msg id record for this recipient, if
+                // possible.
+                let chunks: &mut [A8Bytes<U24>] = buffer.as_mut_aligned_chunks();
+                for (idx, chunk) in chunks.iter_mut().enumerate() {
```

Comment:
Why does chunks need to be mutable in this loop?

Reply:
good catch, thank you